### PR TITLE
feat: assign (message/publisher/passing) function

### PR DESCRIPTION
**Pull request by rokamu623**
`def assign_message_context(self, node_name: str, context_type, sub_topic_name: str, pub_topic_name: str):`

`def assign_publisher(self, node_name: str, pub_topic_name: str, callback_function: Optional[CallbackStruct]):`

`def assign_message_passings(self, node_name: str, source_callback_name: str, destination_callback_name: str):`

is implemented here.

### CODE REVIEW

File: src/test/architecture/test_architecture.py [hunk 1]
```diff
@@ -312,3 +312,64 @@ def test_verify_callback_uniqueness(self, mocker, caplog):
...
+    - context_type: callback_chain
+      subscription_topic_name: /pong
+      publisher_topic_name: /ping
+"""
+
+        mocker.patch('builtins.open', mocker.mock_open(read_data=architecture_text_expected))
+        arch = Architecture('yaml', 'architecture.yaml')
+        node_expected = arch.get_node("/pong_node")
+
+        assert len(node.variable_passings) == len(node_expected.variable_passings)
```
Comment (hsgwa) [inline src/test/architecture/test_architecture.py:374]: VariableObjects can compare themselves.

```suggestion
        assert node.variable_passings == node_expected.variable_passings
        assert node.publishers == node_expected.publishers
```

File: src/caret_analyze/architecture/architecture.py [hunk 2]
```diff
@@ -207,6 +207,33 @@ def _verify(nodes: Collection[NodeStruct]) -> None:
...
                      f'period_ns: {uniqueness_violated[1]}'))
 
+    def assign_message_context(self, node_name: str, context_type, sub_topic_name: str, pub_topic_name: str):
+        try:
+            node: NodeStruct = Util.find_one(lambda x: x.node_name == node_name, self._nodes)
+            node.assign_message_context(context_type, sub_topic_name, pub_topic_name)
+        except ItemNotFoundError:
+            msg = 'Failed to find node. '
+            msg += f'node_name: {node_name}'
+            raise ItemNotFoundError(msg)
```
Comment (hsgwa) [inline src/caret_analyze/architecture/architecture.py:217]: I would like to refactor the ItemNotFoundError exception message in the future, so simplicity is fine here.
```suggestion
        node: NodeStruct = Util.find_one(lambda x: x.node_name == node_name, self._nodes)
        node.assign_message_context(context_type, sub_topic_name, pub_topic_name)
```

File: src/test/architecture/test_architecture.py (hunk 1 above)
Comment (rokamu623) [reply src/test/architecture/test_architecture.py]: Fixed it in [50b4698](https://github.com/tier4/CARET_analyze/pull/196/commits/50b4698dffa5b3c7368e101c3185d1f8b5e47ef0)

File: src/caret_analyze/architecture/architecture.py (hunk 2 above)
Comment (rokamu623) [reply src/caret_analyze/architecture/architecture.py]: Fixed it in [50b4698](https://github.com/tier4/CARET_analyze/pull/196/commits/50b4698dffa5b3c7368e101c3185d1f8b5e47ef0)

File: src/caret_analyze/architecture/struct/node.py [hunk 3]
```diff
@@ -163,3 +164,25 @@ def to_value(self) -> NodeStructValue:
             else tuple(v.to_value() for v in self.callback_groups),
             None if self.variable_passings is None
             else tuple(v.to_value() for v in self.variable_passings))
+
+    def assign_message_context(self, context_type, sub_topic_name: str, pub_topic_name: str):
+        path = self.get_path(sub_topic_name, pub_topic_name)
+        path.message_context =\
+            MessageContextStruct.create_instance(context_type, Dict(),
+                                                 self.node_name, path.subscription,
+                                                 path.publisher, None)
```
Comment (rokamu623) [inline src/caret_analyze/architecture/struct/node.py:173]: ItemNotFoundError has occurred here.
I investigate the cause.

**Review by rokamu623 — COMMENTED** [review]
Implemented
- assign_publisher()
- assign_message_passings()

Implementing
- assign_message_context()

**Comment by rokamu623** [comment]
main (merged [rename_function](https://github.com/tier4/CARET_analyze/pull/156)) is merged to this branch.
[rename_function](https://github.com/tier4/CARET_analyze/pull/156)-compliant implementation is in progress.

**Review by rokamu623 — COMMENTED** [review]
[rename_function](https://github.com/tier4/CARET_analyze/pull/156)-compliant implementation

File: src/caret_analyze/architecture/struct/callback.py [hunk 4]
```diff
@@ -112,6 +112,11 @@ def publish_topic_names(self) -> Optional[List[str]]:
     def to_value(self) -> CallbackStructValue:
         pass
 
+    def assign_publisher(self, pub_topic_name: str):
+        if self._publish_topic_names is not None and\
+           pub_topic_name not in self._publish_topic_names:
+            self._publish_topic_names.append(pub_topic_name)
```
Comment (hsgwa) [inline src/caret_analyze/architecture/struct/callback.py:118]: If none, please add pub_topic_name to a new list.
```suggestion
        self._publish_topic_names = self._publish_topic_names or []
        if pub_topic_name not in self._publish_topic_names:
            self._publish_topic_names.append(pub_topic_name)
```

File: src/caret_analyze/architecture/struct/node.py [hunk 5]
```diff
@@ -164,6 +165,41 @@ def to_value(self) -> NodeStructValue:
...
+            self._node_paths.append(path)
+
+    def assign_publisher(self, pub_topic_name: str,
+                         callback_function: Optional[CallbackStruct]):
+        publisher = PublisherStruct(self.node_name, pub_topic_name,
+                                    [] if callback_function is None else [callback_function])
+        self._publishers.append(publisher)
+
+    def assign_message_passings(self, source_callback: CallbackStruct,
+                                destination_callback: CallbackStruct):
```
Comment (hsgwa) [inline src/caret_analyze/architecture/struct/node.py:195]: It is better to provide an interface that is easy to test.
```suggestion
    def assign_message_passings(self, source_callback: str, destination_callback: str):
```

File: src/caret_analyze/architecture/struct/node.py [hunk 6]
```diff
@@ -164,6 +165,41 @@ def to_value(self) -> NodeStructValue:
...
+            message_context =\
+                MessageContextStruct.create_instance(context_type, {},
+                                                     self.node_name, subscription,
+                                                     publisher, None)
+            path = NodePathStruct(self.node_name, subscription, publisher,
+                                  None, message_context)
+            self._node_paths.append(path)
+
+    def assign_publisher(self, pub_topic_name: str,
+                         callback_function: Optional[CallbackStruct]):
```
Comment (hsgwa) [inline src/caret_analyze/architecture/struct/node.py:189]:
```suggestion
    def assign_publisher(self, pub_topic_name: str, callback_name: Optional[str]):
```

**Review by hsgwa — CHANGES_REQUESTED** [review]
Please make some interface changes.

File: src/caret_analyze/architecture/struct/callback.py (hunk 4 above)
Comment (rokamu623) [reply src/caret_analyze/architecture/struct/callback.py]: Fix it in [93f96b9](https://github.com/tier4/CARET_analyze/pull/196/commits/93f96b9a54029c50c1d9392d5c2748cfb8ab3e3c).

File: src/caret_analyze/architecture/struct/node.py (hunk 5 above)
Comment (rokamu623) [reply src/caret_analyze/architecture/struct/node.py]: Fix it in [93f96b9](https://github.com/tier4/CARET_analyze/pull/196/commits/93f96b9a54029c50c1d9392d5c2748cfb8ab3e3c), and change implement of Archtecture class.

File: src/caret_analyze/architecture/struct/node.py [hunk 7]
```diff
@@ -186,6 +187,48 @@ def to_value(self) -> NodeStructValue:
...
+                MessageContextStruct.create_instance(context_type, {},
+                                                     self.node_name, subscription,
+                                                     publisher, None)
+            path = NodePathStruct(self.node_name, subscription, publisher,
+                                  None, message_context)
+            self._node_paths.append(path)
+
+    def assign_publisher(self, pub_topic_name: str, callback_function_name: str):
+        callback: CallbackStruct = \
+            Util.find_one(lambda x: x.callback_name == callback_function_name, self.callbacks)
```
Comment (rokamu623) [inline src/caret_analyze/architecture/struct/node.py:212]: Change of commit [93f96b9](https://github.com/tier4/CARET_analyze/pull/196/commits/93f96b9a54029c50c1d9392d5c2748cfb8ab3e3c).
Search callback in NodeStruct::assign_publisher not in Architecture::assign_publisher.

File: src/caret_analyze/architecture/struct/node.py [hunk 8]
```diff
@@ -164,6 +165,41 @@ def to_value(self) -> NodeStructValue:
...
+        self._publishers.append(publisher)
+
+    def assign_message_passings(self, source_callback: CallbackStruct,
+                                destination_callback: CallbackStruct):
+        passing = VariablePassingStruct(self.node_name, destination_callback, source_callback)
+        # passing_list = [] if self.variable_passings is None else list(self.variable_passings)
+        if self._variable_passings_info is None:
+            self._variable_passings_info = [passing]
+        else:
+            self._variable_passings_info.append(passing)
```
[the remark this answers is not in the document]
Comment (hsgwa) [reply src/caret_analyze/architecture/struct/node.py:230]: LGTM

File: src/caret_analyze/architecture/struct/node.py (hunk 7 above)
Comment (hsgwa) [reply src/caret_analyze/architecture/struct/node.py:212]: LGTM

File: src/caret_analyze/architecture/struct/node.py [hunk 9]
```diff
@@ -186,6 +187,48 @@ def to_value(self) -> NodeStructValue:
...
+        callback.assign_publisher(pub_topic_name)
+        publisher = PublisherStruct(self.node_name, pub_topic_name, [callback])
+
+        self._publishers.append(publisher)
+
+    def assign_message_passings(self, src_callback_name: str, des_callback_name: str):
+        source_callback =\
+            Util.find_one(lambda x: x.callback_name == src_callback_name, self.callbacks)
+        destination_callback =\
+            Util.find_one(lambda x: x.callback_name == des_callback_name, self.callbacks)
```
[the remark this answers is not in the document]
Comment (hsgwa) [reply src/caret_analyze/architecture/struct/node.py:223]: LGTM

File: src/caret_analyze/architecture/struct/node.py (hunk 3 above)
Comment (rokamu623) [reply src/caret_analyze/architecture/struct/node.py]: Resolve it in [d28b3db](https://github.com/tier4/CARET_analyze/pull/196/commits/d28b3db1e323c231279dd1a53007e8d8f58253bb).
If `ItemNotFoundError` has occurred, new `MessageContextStruct` is created and appended.

https://github.com/tier4/CARET_analyze/blob/bb4ae6b5b7cbef83ce8960979df11e53e4a00481/src/caret_analyze/architecture/struct/node.py#L197-L208

**Review by hsgwa — CHANGES_REQUESTED** [review]
Could you please add error test cases?

**Comment by rokamu623** [comment]
Tests of invalid assignment is implemented in [cfbac79](https://github.com/tier4/CARET_analyze/pull/196/commits/cfbac79f679099ef8d538bdc1c4250f77b2d0d09).
https://github.com/tier4/CARET_analyze/blob/1741fbdf0710af2c6adbc8f28d323970a68af470/src/test/architecture/test_architecture.py#L416-L420

**Review by hsgwa — COMMENTED** [review]
Please add test cases for the duplicate "assign" cases.

**Comment by takam5f2** [comment]
@hsgwa @rokamu623   
I'm investigating the testing failure on `src/test/architecture/test_architecture.py::TestArchitecture::test_assign_publisher`.
I found that the following difference, but I cannot find cause so far.

```
(Pdb) arch.get_node('/pong_node').paths[2]._child == arch_expected.get_node('/pong_node').paths[1]._child
False
```

**Comment by rokamu623** [comment]
rename `assign_message_passing` -> `assign_variable_passing`.

Implementation has been made to resolve the problem of incorrect NodePathStruct structure when `assign_XXX` is executed.
DONE:
- assign_publisher
- assign_variable_passing
WIP:
- assign_message_context

A test is implemented when an incorrect value is entered.
A test is implemented when an duplicate value is entered.
DONE:
- assign_publisher
- assign_variable_passing
WIP:
- assign_message_context

**Comment by rokamu623** [comment]
A test is implemented when an incorrect value is entered.
A test is implemented when an duplicate value is entered.
DONE:
- assign_publisher
- assign_variable_passing

WIP:
- assign_message_context

**Comment by rokamu623** [comment]
Implementation has been made to resolve the problem of incorrect NodePathStruct structure when assign_XXX is executed.
DONE:

- assign_message_context

https://github.com/tier4/CARET_analyze/blob/fd02c3265ee6c245bc0f3def2bda598a21384a6c/src/caret_analyze/architecture/architecture.py#L380-L386

File: src/caret_analyze/architecture/architecture.py [hunk 10]
```diff
@@ -305,6 +306,46 @@ def _verify(nodes: Collection[NodeStruct]) -> None:
...
+                               subscribe_topic_name: str, publish_topic_name: str):
+        node: NodeStruct =\
+            Util.find_one(lambda x: x.node_name == node_name, self._nodes)
+
+        if publish_topic_name not in node.publish_topic_names:
+            raise ItemNotFoundError('{pub_topic_name} is not found in {node_name}')
+
+        if subscribe_topic_name not in node.subscribe_topic_names:
+            raise ItemNotFoundError('{sub_topic_name} is not found in {node_name}')
+
```
Comment (hsgwa) [inline src/caret_analyze/architecture/architecture.py:319]: It seems that these implementations are redundant.

```suggestion

```

**Review by hsgwa — CHANGES_REQUESTED** [review]

File: src/caret_analyze/architecture/architecture.py (hunk 10 above)
Comment (rokamu623) [reply src/caret_analyze/architecture/architecture.py:319]: @hsgwa 
This is an implementation that returns errors when trying to assign a topic that does not exist in `Architecture::NodeStruct`.

From that point of view, this error handling needs to remain but should be replaced with `InvalidArgumentError`.

**Review by hsgwa — APPROVED** [review]
LGTM
